common/src/main/java/com/microsoft/identity/common/internal/controllers/NativeAuthMsalController.kt

```diff
@@ -652,6 +827,274 @@ class NativeAuthMsalController : BaseNativeAuthController() {
+                return performSignUpChallengeCall(
+                    oAuth2Strategy = oAuth2Strategy,
+                    signupToken = this.signupToken
+                ).toSignUpStartCommandResult() as SignUpSubmitCodeCommandResult
+            }
+
+            is SignUpContinueApiResult.InvalidOOBValue -> {
+                SignUpCommandResult.InvalidCode(
+                    error = this.error,
+                    errorDescription = this.errorDescription
+                )
+            }
+            is SignUpContinueApiResult.Redirect -> {
+                INativeAuthCommandResult.Redirect()
+            }
+            is SignUpContinueApiResult.UnknownError -> {
+                Logger.warn(
+                    TAG,
+                    "Unexpected result: $this"
+                )
+                INativeAuthCommandResult.UnknownError(
+                    error = this.error,
+                    errorDescription = this.errorDescription,
+                    details = this.details
+                )
+            }
+
+            is SignUpContinueApiResult.InvalidAttributes, is SignUpContinueApiResult.InvalidPassword -> {
+                Logger.warn(
+                    TAG,
+                    "Unexpected result: $this"
+                )
+                INativeAuthCommandResult.UnknownError(
+                    error = "unexpected_api_result",
+                    errorDescription = "API returned unexpected result: $this"
+                )
+            }
+        }
+    }
+
+    private fun SignUpContinueApiResult.toSignUpSubmitUserAttributesCommandResult(
+        oAuth2Strategy: NativeAuthOAuth2Strategy
+    ): SignUpSubmitUserAttributesCommandResult {
+        return when (this) {
+            is SignUpContinueApiResult.Success -> {
+                SignUpCommandResult.Complete(
+                    signInSLT = this.signInSLT,
+                    expiresIn = this.expiresIn
+                )
+            }
+            is SignUpContinueApiResult.UsernameAlreadyExists -> {
+                SignUpCommandResult.UsernameAlreadyExists(
+                    error = this.error,
+                    errorDescription = this.errorDescription
+                )
+            }
+            is SignUpContinueApiResult.AttributesRequired -> {
+                SignUpCommandResult.AttributesRequired(
+                    signupToken = this.signupToken,
+                    error = this.error,
+                    errorDescription = this.errorDescription,
+                    requiredAttributes = this.requiredAttributes
+                )
+            }
+            is SignUpContinueApiResult.CredentialRequired -> {
+                return performSignUpChallengeCall(
+                    oAuth2Strategy = oAuth2Strategy,
+                    signupToken = this.signupToken
+                ).toSignUpStartCommandResult() as SignUpSubmitUserAttributesCommandResult // TODO can we find something more graceful than a runtime cast?
```

Comment:
Check with Sammy on this one, I think we were working on removing such casts during runtime while I was returning to android MSAL team

Reply:
I left sort of related comments in a few places. I think we should make the return types of methods more explicit rather than generic and then having to do all these casts